consensus/types/src/attester_slashing.rs

```diff
@@ -1,34 +1,114 @@
 use crate::{test_utils::TestRandom, EthSpec, IndexedAttestation};
-
 use derivative::Derivative;
 use serde::{Deserialize, Serialize};
 use ssz_derive::{Decode, Encode};
+use superstruct::superstruct;
 use test_random_derive::TestRandom;
 use tree_hash_derive::TreeHash;
 
-/// Two conflicting attestations.
-///
-/// Spec v0.12.1
+#[superstruct(
+    variants(Base, Electra),
+    variant_attributes(
+        derive(
+            Derivative,
+            Debug,
+            Clone,
+            Serialize,
+            Deserialize,
+            Encode,
+            Decode,
+            TreeHash,
+            TestRandom,
+            arbitrary::Arbitrary
+        ),
+        derivative(PartialEq, Eq, Hash(bound = "E: EthSpec")),
+        serde(bound = "E: EthSpec"),
+        arbitrary(bound = "E: EthSpec")
+    ),
+    ref_attributes(derive(Debug))
+)]
 #[derive(
-    Derivative,
-    Debug,
-    Clone,
-    Serialize,
-    Deserialize,
-    Encode,
-    Decode,
-    TreeHash,
-    TestRandom,
-    arbitrary::Arbitrary,
+    Debug, Clone, Serialize, Encode, Deserialize, TreeHash, Derivative, arbitrary::Arbitrary,
 )]
 #[derivative(PartialEq, Eq, Hash(bound = "E: EthSpec"))]
-#[serde(bound = "E: EthSpec")]
+#[serde(bound = "E: EthSpec", untagged)]
 #[arbitrary(bound = "E: EthSpec")]
+#[ssz(enum_behaviour = "transparent")]
+#[tree_hash(enum_behaviour = "transparent")]
 pub struct AttesterSlashing<E: EthSpec> {
     pub attestation_1: IndexedAttestation<E>,
     pub attestation_2: IndexedAttestation<E>,
 }
 
+/// This is a copy of the `AttesterSlashing` enum but with `Encode` and `Decode` derived
+/// using the `union` behavior for the purposes of persistence on disk. We use a separate
+/// type so that we don't accidentally use this non-spec encoding in consensus objects.
+#[derive(Debug, Clone, Encode, Decode, Derivative)]
+#[derivative(PartialEq, Eq, Hash(bound = "E: EthSpec"))]
+#[ssz(enum_behaviour = "union")]
+pub enum AttesterSlashingOnDisk<E: EthSpec> {
+    Base(AttesterSlashingBase<E>),
+    Electra(AttesterSlashingElectra<E>),
+}
```

Comment:
This is a breaking change to the on-disk format, because we add the tag. I would ideally like to release v5.2.0 _without_ a schema upgrade, so that if anything goes wrong with tree-states, users are able to downgrade easily.

However this is not a hard requirement, so we have a few options:

1) Do the database upgrade and include in v5.2.0
2) Wait to merge this PR until after v5.2.0 (maybe not a good option with interop fast approaching, but feasible if we fast-track v5.2.0)
3) Merge this PR with a temporary implementation of `Decode` which tries both variants in order. This is obviously janky and not ideal, so the plan would be to clean it up in a subsequent PR.

My preference is for (2), then (3), then (1)

Reply:
I'm in favor of fast tracking 5.2.0 because there's gonna be a *ton* of changes in Electra and it's probably best to separate all the other changes that have been merged in from the Electra code.

Reply:
Note that you can infer the tag with the current database schema. A phase0 `AttesterSlashing` is:

```python
class AttesterSlashing(Container):
    attestation_1: IndexedAttestation
    attestation_2: IndexedAttestation
```

Where `IndexedAttestation` is a variable size type. So the first 8 bytes of the serialized stream are offset. The first offset will always be `00000008` and the second is variable on the size of the first attestation. So a non-tag `AttesterSlashing` always starts by `00000008` or just `00` and a tagged `AttesterSlashing` can start by `0100000008` or just `01`. Now for `00` the deserializer should consider the tag size to process the rest of the stream, or append a single zero byte

Reply:
@dapplion I'm fine with that kind of implementation so long as it's well-tested. We don't have any machinery to generate code like that, so it will need to be carefully written by hand and kept up to date.